velox/functions/prestosql/types/JsonType.cpp

```diff
@@ -168,9 +169,37 @@ struct AsJson {
       const SelectivityVector& rows,
       const BufferPtr& elementToTopLevelRows,
       bool isMapKey = false)
-      : decoded_(context, *input, rows) {
+      : decoded_(context) {
     ErrorVectorPtr oldErrors;
     context.swapErrors(oldErrors);
+    if (isJsonType(input->type())) {
+      json_ = input;
+    } else {
+      if (!exec::PeeledEncoding::isPeelable(input->encoding()) ||
```

Comment:
That was my initial attempt as well, but it looks like this class in specific is being used as a helper class (which keeps a decoded vector as live state) to generate a json and I found it difficult to connect the code paths from cast to this. You might have more context on the Json related code so I am open to suggestions if you find there is a better way as combining this would be great if we can.